src/main/protobuf/remote_output_service.proto

```diff
@@ -0,0 +1,320 @@
+  // file, they are replaced by an empty directory. If a file or
+  // directory already exists at the provided path, it is replaced.
+  //
+  // This means that symbolic links are not followed when evaluating
+  // path_prefix and OutputFile.path.
+  repeated build.bazel.remote.execution.v2.OutputFile files = 4;
+
+  // Symbolic links that need to be created.
+  //
+  // Any missing parent directories, including those in path_prefix, are
+  // created as well. If any of the parents refer to a non-directory
+  // file, they are replaced by an empty directory. If a file or
+  // directory already exists at the provided path, it is replaced.
+  //
+  // This means that symbolic links are not followed when evaluating
+  // path_prefix and OutputSymlink.path.
+  repeated build.bazel.remote.execution.v2.OutputSymlink symlinks = 5;
+
+  // Directories that need to be downloaded from the Content Addressable
+  // Storage.
+  //
+  // Any missing parent directories, including those in path_prefix, are
+  // created as well. If any of the parents refer to a non-directory
+  // file, they are replaced by an empty directory. Any file or
+  // directory that already exists at the provided path is replaced.
+  //
+  // This means that symbolic links are not followed when evaluating
+  // path_prefix and OutputDirectory.path.
+  repeated build.bazel.remote.execution.v2.OutputDirectory directories = 6;
+}
+
+message BatchStatRequest {
+  // The identifier of the build. The remote output service uses this to
+  // determine which output path needs to be inspected.
+  string build_id = 1;
+
+  // In case the path corresponds to a regular file, include the hash
+  // and size of the file in the response.
+  bool include_file_digest = 2;
+
+  // In case the path corresponds to a symbolic link, include the target
+  // of the symbolic link in the response.
+  bool include_symlink_target = 3;
+
+  // If the last component of the path corresponds to a symbolic link,
+  // return the status of the file at the target location.
+  //
+  // Symbolic links encountered before the last component of the path
+  // are always expanded, regardless of the value of this option.
+  bool follow_symlinks = 4;
+
+  // Paths whose status needs to be obtained.
+  repeated string paths = 5;
+}
+
+message BatchStatResponse {
+  // The status response for each of the requested paths, using the same
+  // order as requested. This means that this list has the same length
+  // as BatchStatRequest.paths.
+  repeated StatResponse responses = 1;
+}
+
+message StatResponse {
+  // The status of the file. If the file corresponding with the
+  // requested path does not exist, this field will be null.
+  FileStatus file_status = 1;
+}
+
+message FileStatus {
+  message File {
+    // The hash and size of the file. This field is only set when
+    // BatchStatRequest.include_file_digest is set.
+    build.bazel.remote.execution.v2.Digest digest = 1;
```

Comment:
Some more experience using this protocol: my recommendation would be to allow the server to not report this value if it prefers.

Namely, I have observed Bazel sometimes prefers to obtain the digest of a file for which it still has one or more open file descriptors. If the FUSE file system has write-back caching enabled, this may cause dirty pages to remain in the page cache.

In that case it would be desirable for bb_clientd or any other implementation to omit this field, instructing Bazel to compute the digest itself.

```proto
     // The hash and size of the file. This field is only set when
     // BatchStatRequest.include_file_digest is set.
+    //
+    // This field may also be omitted if the remote output service is
+    // unable to compute it accurately. For example, when a file is
+    // opened for writing, the kernel may buffer data to be written.
+    // When absent, the caller should fall back to computing the digest
+    // manually.
     build.bazel.remote.execution.v2.Digest digest = 1;
```

Reply:
I agree. From Bazel's perspective, this is a best-effort/fast approach to get digest if available. Bazel can always fallback to manually compute the digest.

Reply:
Good :) I will update the comment